### ISSUE TITLE
Feature Request: PSBT creation should include unspentTargetCount

### ISSUE THREAD

**Issue by junderw** [dgarage/NBXplorer#183]
`unspentTargetCount` Optional. When passed, the wallet will attempt to make sure the UTXO count of the wallet after this PSBT is signed and confirmed is equal to this number.

----

The use case for this:

* In a wallet that periodically sends out transactions, keeping about 3-4 UTXOS in your wallet at all times makes it easy to ensure that one stuck transaction does not cause your wallet to choke up on the 25 unconfirmed tx chain limit... As it is unlikely that all 3 or all 4 of the "base" (confirmed) UTXOs will all accidentally be sent with low fees and get stuck.

----

Implementation ideas:

- Simple: make this a "max" setting... then all you need to do is add some low value outpoints into the transaction as inputs to join everything together decreasing UTXO count to desired amount.
- Complex: Have min and mix... is trickier, because if UTXO count is too low, you must maybe create multiple change outputs to increase UTXO count, which then brings up many extra questions about how to handle explicitChangeAddress etc. etc...

**Comment by NicolasDorier** [comment]
Can you reformulate? I don't understand.

**Comment by junderw** [comment]
ok, ignore the reason and use case I will only explain the logic of the feature:

1. I set it to 3
2. My wallet has 3 UTXOs right now, 1, 2, and 3 BTC
3. I send 1 BTC, it uses one utxo, and creates one change. The result is 3 utxo in my wallet, I am happy.
4. I send 4BTC, it needs to spend two utxo, so if I only have one change output, I will have only 2 utxo in my wallet. This feature will instead create 2 change outputs instead of 1, so my wallet will end up with 3 utxo.
5. Let's say I "refill" my wallet, and forget the amount, so I send 3 txes to refill because I am dumb. now I have 6 utxo. So next time I make PSBT, it will add more inputs and make one big large change output, so maybe I want to only send 0.1, and 1 utxo is enough, but my wallet has 6 utxo and this feature wants a state of 3 utxos. so it spends 4 utxos and creates one large change output, giving me three utxo after confirms.

do you understand the way it operates?

**Comment by junderw** [comment]
We can discuss this in person later. I think this idea needs to be explained with drawings.

Also, BitGo has this feature:

https://www.bitgo.com/api/v2/#operation/v2.wallet.tx.build

The parameter is called `targetWalletUnspents`

**Comment by NicolasDorier** [comment]
Very interesting. I need to think about how to do this.

**Comment by NicolasDorier** [comment]
we also need a `minUTXO` so the change don't get spread too thin.

**Comment by junderw** [comment]
yeah... that way if you divide up the satoshis to the N number of target UTXOs and the size of one UTXO is below the min value, then try targeting N - 1, then N - 2 etc. all the way to 1... If 1 change output is too small........... add it to fee?

**Comment by junderw** [comment]
`target_utxo_count = -1` should mean "don't care about count" and should behave as currently.

**Comment by NicolasDorier** [comment]
NBXplorer already add too small UTXO to the fee. (what is below dust)
Here it would be a different minimum, above the dust be below some target value that make sense for the business.

**Comment by NicolasDorier** [comment]
In the NBX API this will be an optional parameter, if not specified or `null`, I will just keep current behavior.

**Comment by junderw** [comment]
I see.

So if even 1 change is below the value, then just make the 1 change output anyways.

**Comment by NicolasDorier** [comment]
yes
